### ISSUE TITLE
handle flags for argo

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#896]
This rests on #892, although it seems likely that working on both at the same time will prove to be useful.

**Comment by richardsc** [comment]
This is coming along in my fork, and I'll merge with develop once @dankelley has updated with all the documentation changes from the `kelley` branch.

**Comment by richardsc** [comment]
PR #910 brings in my additions to have an `argo-method` for `handleFlags()`. Things seem to be working, except there are some differences in behaviour with the examples in the `ctd-method` that I can't reconcile. A reproducible example is below:

From the `?handleFlags,ctd-method` example number 2:
```r
library(oce)
data(section)
stn <- section[["station", 100]]
STN <- handleFlags(stn, flags=list(salinity=c(1, 4:9)))
```
Note that this does not require the `actions` arg to be specified -- I guess it is assumed to be "NA".

Now let's try the same thing with the new `handleFlags,argo-method`:
```r
data(argo)
argoNew <- handleFlags(argo, flags=list(salinity=4))
```
```
 Error in handleFlagsInternal(object, flags, actions) : 
  action must be a character string or a function 
```
@dankelley I'd appreciate if you could have a look at what's going on here.

**Comment by dankelley** [comment]
sure -- will be easy. Tonight sometime.

**Comment by dankelley** [comment]
The roxygen work for read.adp was a lot tricker than I thought it would be, so I'm a bit too tired to look at this. But I just assigned myself, and I will look on Wed morning or afternoon. I think I know where the problem must be, but I'd like to not just fix the bug but also maybe to play a bit more to see if these things  "feel" okay. Also, I might get some insights if I can find time to go over to that R/dfo meeting tomorrow (seems unlikely given calendar fillup).

**Comment by richardsc** [comment]
In case it helps, I've been doing some debugging, and it seems like the problem is likely around line [398-399](https://github.com/dankelley/oce/blob/develop/R/AllClass.R#L399) of `AllClass.R` -- at the spot when `actionsThis` is determined:
```r
actionsThis <- if (all) actions[[1]] else actions[[name]]
```
Because `flags` was passed as a list with `salinity`, the variable `all` is `FALSE`, and the result is `actionsThis` gets set to `NULL`, because `actions[[name]]` doesn't exist:
```r
Browse[1]> actions
[[1]]
[1] "NA"

Browse[1]> flagsThis
[1] 1
Browse[1]> flags
$salinity
[1] 1

Browse[1]> actionsThis
NULL
Browse[1]> actions[[name]]
NULL
```

What I don't understand is why this works (or seems to work?) for the CTD case and not here ...

**Comment by dankelley** [comment]
I've set up something in the oce-issues repo on this, in [896a.R](https://github.com/dankelley/oce-issues/blob/master/896/896a.R). I don't think it's doing the ctd case right, either. I'll look into it.

**Comment by dankelley** [comment]
I think this (https://github.com/dankelley/oce/issues/896#issuecomment-206053714) is ok now, in `develop` commit ce32f728d785f2409ef5f1ec1e5fb0c491517626.

Importantly, I've added tests to the test suite. I like to do that so we can ensure that (a) things are ok now and (b) these particular things cannot break in the future without us knowing about it.

@richardsc -- I figure we might want to keep this issue open for a while, as we gain some experience in how the UI should be set up.  (And by "we" I probably mean "you"!) Feel free to close any time, or leave open until we both like the UI.

**Comment by richardsc** [comment]
This is looking good. I'll assign myself, and will close after some time that seems appropriate depending on how the flags work is going.

**Comment by richardsc** [comment]
One thing I just thought of is that `handleFlags()` should be adding an entry to the `processingLog()`. I'm not sure if it's best to do this in the class specific function, or in `handleFlagsInternal()`.

Any thoughts @dankelley ?

**Comment by dankelley** [comment]
I agree. I think it should go in the class-specific, because we want to say some details (e.g. that WOCE type was used, or that argo was used as per such-such documentation). This will mean more coding this way, but otherwise the log won't mean all that much to people.  I am thinking of a two-step thing, e.g. the class specific could say `"flags and actions set by user"` or `"flags set by user and actions taken as default"` or maybe `"flags set by default for WOCE data"`. I'm glad we are starting with just 2 classes ... it will make it a lot easier to try things and see how they look in the code and to the user.

I'll try something for ctd, maybe this afternoon.

**Comment by dankelley** [comment]
I wrote something about CTD flags in the vignette, but it was orientated to *manipulating* flags, not using them for an action. If you're looking for a little task to see if the argo flag system "feels right", I think the vignette might be a good place to write something.  (I find that the necessity of writing sentences clarifies thoughts ... for example, I had never actually changed a flag before I wrote that vignette entry, although in fact I think changing flags is an important application for e.g. DFO users.)

**Comment by dankelley** [comment]
I wrote a blog posting on handling section flags (http://dankelley.github.io/r/2016/04/10/flags.html) and I realized ... that I wasn't handling them!  I also found about 3 bugs, doing that. So, I encourage Clark to write a blog posting for argo, since that might also uncover some bugs that we don't want to pester us when we are using these new features in real work.

**Comment by richardsc** [comment]
Nice post! I like the two contour plots at the end that show the full data and the "good" data -- where the latter is missing the little odd-looking contoured "peaks".

I plan to write a little post about argo and flags. In fact, I found a neat dataset that uses a variety of flags, that has some interesting looking signals, that I think would be fun to explore. I'll keep this issue open until I've managed to do this.
